File: erp-backend/src/main/java/com/erp/controller/ExpenseController.java
package com.erp.controller;

import com.erp.entity.Expense;
import com.erp.repository.ExpenseRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.data.domain.Sort;

import java.math.BigDecimal;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@RestController
public class ExpenseController {

    @Autowired
    private ExpenseRepository expenseRepository;

    // POST: Naya kharcha (Expense) add karna
    @PostMapping("/add-expense")
    public ResponseEntity<?> addExpense(@RequestAttribute("role") String role, @RequestBody Map<String, Object> body) {
        if (!"admin".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied: Only Admin can add expenses");
        }
        Expense expense = new Expense();
        expense.setDescription((String) body.get("description"));
        expense.setAmount(new BigDecimal(body.get("amount").toString()));
        expense.setCategory((String) body.get("category"));

        Expense saved = expenseRepository.save(expense);

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("message", "Expense added!");
        response.put("data", saved);
        return ResponseEntity.ok(response);
    }

    // GET: Tamam kharche (Expenses) list get karna
    @GetMapping("/expenses")
    public ResponseEntity<?> getAllExpenses(@RequestAttribute("role") String role) {
        if (!"admin".equalsIgnoreCase(role)) {
            return ResponseEntity.status(403).body("Access Denied: Only Admin can view expenses");
        }
        List<Expense> expenses = expenseRepository.findAll(Sort.by(Sort.Direction.DESC, "expenseId"));
        return ResponseEntity.ok(expenses);
    }
}

File: erp-backend-node/src/controllers/expenseController.ts
import { Request, Response } from 'express';
import prisma from '../prisma/client';

// Get all expenses
export const getExpenses = async (req: Request, res: Response) => {
  try {
    const expenses = await prisma.expenses.findMany({
      orderBy: { date: 'desc' }
    });
    res.json(expenses);
  } catch (error) {
    console.error('Error fetching expenses:', error);
    res.status(500).json({ message: 'Server error while fetching expenses.' });
  }
};

// Add a new expense
export const addExpense = async (req: Request, res: Response) => {
  try {
    const { description, amount, category } = req.body;
    
    const newExpense = await prisma.expenses.create({
      data: {
        description,
        amount,
        category,
        date: new Date()
      }
    });
    res.json(newExpense);
  } catch (error) {
    console.error('Error adding expense:', error);
    res.status(500).json({ message: 'Server error while adding expense.' });
  }
};
